---
title: "Example Isoform Plots in Target Clusters"
author: "Mervin M Fansler"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
---

# Purpose

Exploring visualization of isoform-level expression across perturbations.

# Initialization

## Libraries
```{r libs, message=FALSE, warning=FALSE}
library(magrittr)
library(tidyverse)
library(Matrix)
library(matrixStats)
library(SingleCellExperiment)
library(Gviz)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(plyranges)
```

## Parameters
```{r set_params}
set.seed(20210818)

options(ucscChromosomeNames=FALSE)
BIN_WIDTH=50
N_EXAMPLES=5


txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

FILE_SE="data/se/kd6_essential_bulk_expressed.Rds"
FILE_GTF="../hcl/data/gff/utrome.e30.t5.gc39.pas3.f0.9999.w500.gtf.gz"
FILE_CLUSTERS="tbl/df_target_nnclusters_kd6_essential_ui10.csv"
FILE_GENE_CLUSTERS="tbl/df_gene_nnclusters_kd6_essential_ui10.csv"

LBL_UTR_CLUSTERS=c(
    "14"="CFIm complex",
    "11"="mRNA export",
    "13"="proteosome",
    "9"="mitotic cell cycle",
    "16"="chromatin organization",
    "18"="chaperonin-containing t-complex",
    "15"="cytoplasmic translation",
    "2"="ribosome",
    "12"="ribosome biogenesis",
    "5"="translation initiation",
    "1"="mRNA splicing I",
    "6"="mRNA splicing II",
    "8"="transcription initiation",
    "3"="tRNA activation",
    "7"="DNA replication and repair",
    "4"="transcription initiation and elongation",
    "10"="CPSF, CSTF, and PAF complex",
    "17"="nuclear exosome complex",
    "0"="non-targeting"
)
```


# Data
## Loading
```{r load_data, message=FALSE}
se <- readRDS(FILE_SE)
gr <- read_gff(FILE_GTF, genome_info='hg38') %>%
    filter(type == 'transcript') %>%
    anchor_3p() %>%
    mutate(width=BIN_WIDTH)
df_clusters <- read_csv(FILE_CLUSTERS)
df_gene_clusters <- read_csv(FILE_GENE_CLUSTERS, col_types='cci')
tr_genome <- GenomeAxisTrack(v=10)
```

## Preprocessing
```{r prepare_data}
gene2txs <- rowData(se) %>% 
    as_tibble %>% 
    dplyr::select(gene_name, transcript_id) %>%
    group_by(gene_name) %>%
    summarize(txs=list(transcript_id)) %>%
    deframe()

tpm_tx_target <- assay(se, "counts") %>% { 
    idx_ntp <- str_detect(colnames(.), "non-targeting")
    cts_ntp <- rowSums(.[,idx_ntp])
    cts_tx_target <- .[, !idx_ntp]
    cbind(cts_tx_target, "ntp"=cts_ntp)
} %>%
    { . %*% Diagonal(ncol(.), 1e6/colSums(.)) }

M_target_cluster <- df_clusters %>% 
    # dummy entry for NTPs
    add_row(sgID_AB="ntp", cluster_id=0) %>%
    dplyr::select(sgID_AB, cluster_id) %>%
    deframe %>% fac2sparse %>% t

tpm_tx_cluster <- M_target_cluster %>%
    { tpm_tx_target[,rownames(.)] %*% . %*% Diagonal(ncol(.), 1/colSums(.)) }
```

## Functions
```{r methods}
get_gene_gr <- function (gene) {
    filter(gr, transcript_id %in% gene2txs[[gene]])
}

get_counts_gr <- function (gene, cluster) {
    get_gene_gr(gene) %>%
        mutate(score=tpm_tx_cluster[transcript_id, cluster]) %>%
        plyranges::select(score)
}

get_tpm_tr <- function (gene, cluster) {
    DataTrack(get_counts_gr(gene, cluster), name=cluster, genome="hg38",
                    type="hist", width=50)
}

get_tpm_tr_all <- function (gene) {
    lapply(names(LBL_UTR_CLUSTERS), get_tpm_tr, gene=gene)
}

get_gene_tr <- function (gene, chr, start, end) {
    GeneRegionTrack(txdb, chromosome=chr, start=start, end=end, symbol=gene,
                    showId=TRUE, name=gene, geneSymbol=TRUE)
}

harmonize_ylim <- function (trs) {
    max_score <- max(sapply(trs, function (tr) { max(tr@data) }))
    ylims <- c(0, max_score*1.05)
    lapply(trs, function (tr) {
        displayPars(tr) <- list(ylim=ylims)
        tr
    })
}

plot_all_clusters <- function (gene) {
    gr_gene <- get_gene_gr(gene)
    chr_gene <- seqnames(gr_gene) %>% as.character %>% `[[`(1)
    start_gene <- min(start(gr_gene)) - 1e3
    end_gene <- max(end(gr_gene)) + 1e3
    
    tr_gene <- get_gene_tr(gene, chr_gene, start_gene, end_gene)

    trs_clusters <- get_tpm_tr_all(gene)
    
    plotTracks(append(harmonize_ylim(trs_clusters), list(tr_gene, tr_genome)),
               #sizes=c(4,4,4,2,2),
               from=start_gene, to=end_gene, 
               col.border.title="white",
               fontcolor.title='grey20',
               col.axis="grey20",
               background.title='grey90')
}

sample_clusters <- function (K) {
    df_gene_clusters %>%
        filter(cluster_id == K) %>%
        slice_sample(n=N_EXAMPLES) %$%
        gene_name
}
```


# Analysis

## Gene Clusters
### Cluster 1
```{r plt_c1, fig.width=4, fig.height=12}
for (gene in sample_clusters(1)) {
    plot_all_clusters(gene)
}
```

### Cluster 2
```{r plt_c2, fig.width=4, fig.height=12}
for (gene in sample_clusters(2)) {
    plot_all_clusters(gene)
}
```

### Cluster 3
```{r plt_c3, fig.width=4, fig.height=12}
for (gene in sample_clusters(3)) {
    plot_all_clusters(gene)
}
```

### Cluster 4
```{r plt_c4, fig.width=4, fig.height=12}
for (gene in sample_clusters(4)) {
    plot_all_clusters(gene)
}
```

### Cluster 5
```{r plt_c5, fig.width=4, fig.height=12}
for (gene in sample_clusters(5)) {
    plot_all_clusters(gene)
}
```

### Cluster 6
```{r plt_c6, fig.width=4, fig.height=12}
for (gene in sample_clusters(6)) {
    plot_all_clusters(gene)
}
```

### Cluster 7
```{r plt_c7, fig.width=4, fig.height=12}
for (gene in sample_clusters(7)) {
    plot_all_clusters(gene)
}
```

### Cluster 8
```{r plt_c8, fig.width=4, fig.height=12}
for (gene in sample_clusters(8)) {
    plot_all_clusters(gene)
}
```

### Cluster 9
```{r plt_c9, fig.width=4, fig.height=12}
for (gene in sample_clusters(9)) {
    plot_all_clusters(gene)
}
```

### Cluster 10
```{r plt_c10, fig.width=4, fig.height=12}
for (gene in sample_clusters(10)) {
    plot_all_clusters(gene)
}
```

### Cluster 11
```{r plt_c11, fig.width=4, fig.height=12}
for (gene in sample_clusters(11)) {
    plot_all_clusters(gene)
}
```

### Cluster 12
```{r plt_c12, fig.width=4, fig.height=12}
for (gene in sample_clusters(12)) {
    plot_all_clusters(gene)
}
```

### Cluster 13
```{r plt_c13, fig.width=4, fig.height=12}
for (gene in sample_clusters(13)) {
    plot_all_clusters(gene)
}
```

### Cluster 14
```{r plt_c14, fig.width=4, fig.height=12}
for (gene in sample_clusters(14)) {
    plot_all_clusters(gene)
}
```

### Cluster 15
```{r plt_c15, fig.width=4, fig.height=12}
for (gene in sample_clusters(15)) {
    plot_all_clusters(gene)
}
```

### Cluster 16
```{r plt_c16, fig.width=4, fig.height=12}
for (gene in sample_clusters(16)) {
    plot_all_clusters(gene)
}
```

### Cluster 17
```{r plt_c17, fig.width=4, fig.height=12}
for (gene in sample_clusters(17)) {
    plot_all_clusters(gene)
}
```

### Cluster 18
```{r plt_c18, fig.width=4, fig.height=12}
for (gene in sample_clusters(18)) {
    plot_all_clusters(gene)
}
```

### Cluster 19
```{r plt_c19, fig.width=4, fig.height=12}
for (gene in sample_clusters(19)) {
    plot_all_clusters(gene)
}
```

---

# Runtime Details
## Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```

## Conda Environment
```{bash comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
