---
title: "Initial QC on RD7 Essential Data"
author: "Mervin M Fansler"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
---

# Purpose

Here we perform a first pass look at the data. We will plot some high-level statistics
and see where the various perturbations fall with respect to these readouts. We 
expect to see that for all readouts, the non-targeting experiments display a central
distribution, with most perturbations falling nearby. Perturbations that impact the
statistics are expected to appear as outliers.

# Initialization

## Libraries
```{r libs, message=FALSE, warning=FALSE}
library(magrittr)
library(tidyverse)
library(plotly)
library(ggbeeswarm)
library(Matrix)
library(matrixStats)
library(SingleCellExperiment)
```

## Parameters
```{r set_params}
set.seed(20210818)
FILE_SCE_TX="data/sce/rd7_essential.annot.txs.Rds"

## KNOWN FACTORS
GENES_CPSF <- c("CPSF1", "CPSF2", "CPSF3", "CPSF4", "WDR33", "FIP1L1")
GENES_CF1 <- c("NUDT21", "CPSF6", "CPSF7")
GENES_CF2 <- c("PCF11", "CLP1")
GENES_CSTF <- c("CSTF1", "CSTF2", "CSTF2T", "CSTF3")
GENES_PAF <- c("PAF1", "CTR9", "RTF1")
GENES_NEF <- c("NXF1", "THOC1", "THOC2", "THOC3", "THOC5", "THOC6", "THOC7")
GENES_SPL <- str_c("SRSF", 1:11)
GENES_MTOR <- c("MTOR", "RPTOR", "MLST8", "CLK2", "AKT1")

COLOR_MAP <- c("non-targeting"="grey",
               "CPSF complex"="red4",
               "CFI complex"="purple2",
               "CFII complex"="orange2",
               "CSTF complex"="peachpuff",
               "PAF complex"="steelblue",
               "mRNA export factors"="orchid2",
               "splicing factors"="seagreen1",
               "MTOR complex"="cyan",
               "other"="black")
```

# Data
## Loading
```{r load_data, message=FALSE}
sce <- readRDS(FILE_SCE_TX)
```

## Preprocessing
Among our statistics, we will compute the percentage of UMIs corresponding to specific
isoform classes, including IPA, proximal, and distal, as well as UMIs from mitochondrial genes.

```{r prepare_data}
idx_mt <- rowRanges(sce) %>% 
    unlist %>% as_tibble %>%
    mutate(is_mt=seqnames == "chrM")  %>% 
    select(transcript_id, is_mt) %>%
    deframe() %>%
    `[`(rownames(sce))

df_cells <- colData(sce) %>% as_tibble %>%
    mutate(umis_total=colSums(counts(sce)),
           pct_ipa=colSums(counts(sce[rowData(sce)$is_ipa,]))/umis_total,
           pct_proximal=colSums(counts(sce[rowData(sce)$is_proximal,]))/umis_total,
           pct_distal=colSums(counts(sce[rowData(sce)$is_distal,]))/umis_total,
           pct_novel=colSums(counts(sce[rowData(sce)$is_novel,]))/umis_total,
           pct_mt=colSums(counts(sce[idx_mt,]))/umis_total)
```

# Analysis
## UMIs per Cell

```{r plt_umis, fig.width=4, fig.height=4}
df_cells %>%
    ggplot(aes(x=UMI_count, y=umis_total)) +
    geom_density2d_filled() +
    geom_abline(linetype='dashed', color="white") +
    scale_x_log10(expand=c(0,0)) + scale_y_log10(expand=c(0,0)) +
    scale_fill_viridis_d(option="F") +
    theme_bw() +
    labs(x="UMI counts per cell [published]", 
         y="UMI counts per cell [scUTRquant]") +
    guides(fill='none')
```

## Cells per Sample
The scUTRquant QC showed the elbow plots were not well defined for Lanes 21 and 48, 
which led to very high numbers of barcodes being called as cells. We hope to see that
these samples do not have a high number of annotation cells.

```{r plt_samples, fig.width=4, fig.height=6}
df_cells %>%
    mutate(sample_id=fct_infreq(sample_id)) %>%
    ggplot(aes(y=sample_id)) +
    geom_bar(fill='lightgrey', color='black', size=0.2) +
    scale_x_continuous(expand=c(0,0,0.05,0)) +
    theme_bw() +
    labs(x="Cells", y="Sample")
```

## Cells per target
```{r plt_targets, fig.width=4, fig.height=3}
df_cells %>%
    mutate(target_gene=fct_infreq(target_gene)) %>%
    dplyr::count(target_gene, name="n_cells") %>%
    filter(target_gene != 'non-targeting') %>%
    ggplot(aes(x=target_gene, y=n_cells)) +
    geom_bar(stat='identity', fill='grey', width=1) +
    scale_y_log10(expand=c(0,0,0.05,0)) +
    theme_bw() +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) + 
    labs(x="Target Gene", y="Cells")
```

There are several perturbations that have extremely low cell counts. Generally, we 
have used a rule of thumb of at least 200 cells per condition as a lower bound for 
performing transcriptome-wide differential transcript usage testing (with scUTRboot).

Instead, for the majority of perturbations, we may only be able to characterize
high-level changes in transcriptome characteristics rather than at gene-level resolution.

# Target Analysis
We will explore whether any perturbations have effects on basic transcriptome characteristics.
These will also be compared against each other.

```{r compute_target_stats}
df_targets <- df_cells %>%
    group_by(target_gene, target_gene_id, sgID_AB) %>%
    summarize(n_cells=dplyr::n(),
              mean_umis_total=mean(umis_total),
              pct_ipa=weighted.mean(pct_ipa, umis_total),
              pct_proximal=weighted.mean(pct_proximal, umis_total),
              pct_distal=weighted.mean(pct_distal, umis_total),
              pct_mt=weighted.mean(pct_mt, umis_total), .groups='drop') %>%
    mutate(gene_set=case_when(
        target_gene == "non-targeting" ~ "non-targeting",
        target_gene %in% GENES_CPSF ~ "CPSF complex",
        target_gene %in% GENES_CF1 ~ "CFI complex",
        target_gene %in% GENES_CF2 ~ "CFII complex",
        target_gene %in% GENES_CSTF ~ "CSTF complex",
        target_gene %in% GENES_PAF ~ "PAF complex",
        target_gene %in% GENES_NEF ~ "mRNA export factors",
        target_gene %in% GENES_SPL ~ "splicing factors",
        target_gene %in% GENES_MTOR ~ "MTOR complex",
        TRUE ~ "other"
    ) %>% factor(levels=c("other", "non-targeting", "CPSF complex", 
                          "CFI complex", "CFII complex", "CSTF complex",
                          "PAF complex", "mRNA export factors",
                          "splicing factors", "MTOR complex"))) %>%
    mutate(is_known=!gene_set %in% c("non-targeting", "other"))
```

### Mitochondrial Extremes
```{r mt_min_max}
df_targets %>%
    select(target_gene, pct_mt, mean_umis_total, sgID_AB) %>%
    slice_max(pct_mt, n=20)

df_targets %>%
    select(target_gene, pct_mt, mean_umis_total, sgID_AB) %>%
    slice_min(pct_mt, n=20)
```

### IPA Extremes
```{r ipa_min_max}
df_targets %>%
    select(target_gene, pct_ipa, mean_umis_total, sgID_AB) %>%
    slice_max(pct_ipa, n=20)

df_targets %>%
    select(target_gene, pct_ipa, mean_umis_total, sgID_AB) %>%
    slice_min(pct_ipa, n=20)
```

### Proximal Extremes
```{r proximal_min_max}
df_targets %>%
    select(target_gene, pct_proximal, mean_umis_total, sgID_AB) %>%
    slice_max(pct_proximal, n=20)

df_targets %>%
    select(target_gene, pct_proximal, mean_umis_total, sgID_AB) %>%
    slice_min(pct_proximal, n=20)
```

### Distal Extremes
```{r distal_min_max}
df_targets %>%
    select(target_gene, pct_distal, mean_umis_total, sgID_AB) %>%
    slice_max(pct_distal, n=20)

df_targets %>%
    select(target_gene, pct_distal, mean_umis_total, sgID_AB) %>%
    slice_min(pct_distal, n=20)
```

## Pair Plots
### Proximal vs Distal
```{r plt_pair_pr_ds, fig.width=6, fig.height=6}
g <- df_targets %>%
    ggplot(aes(x=pct_proximal, y=pct_distal, color=gene_set, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x="Percent UMIs from proximal isoforms",
         y="Percent UMIs from distal isoforms",
         color="Gene Set")
    
ggplotly(g, tooltip=c("text", "x", "y", "gene_set", "n_cells"))
```


```{r plt_pair_pr_ds_static, fig.width=6, fig.height=6}
df_targets %>%
    ggplot(aes(x=pct_proximal, y=pct_distal, color=gene_set, text=target_gene)) +
    geom_point(data=filter(df_targets, gene_set == 'other'), size=0.5) +
    geom_point(data=filter(df_targets, gene_set == 'non-targeting'), size=0.5) +
    geom_point(data=filter(df_targets, is_known), size=3) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    labs(x="Percent UMIs from proximal isoforms",
         y="Percent UMIs from distal isoforms",
         color="Gene Set")
```

Naively, we would expect these to compensatory statistics, however, in this first
pass we include also single-UTR isoforms, which are labeled as both "proximal" and 
"distal". This possibly explains the strong positive correlation.

We observe here three major outliers that 

- NUDT21
- CPSF6
- OGFOD1

The first two are major cleavage enhancement factors that co-complex and bind RNA 
around UGUA motifs found 40-80 nts upstream of the mRNA cleavage site. OGFOD1 has 
no known regulatory effect on 3' UTR isoform usage in mammals, though it is a 
homolog of Tpa1 (*Sac. cer.*), which stands for "Termination and Polyadenylation 1".

This could represent novel function.

### IPA vs Proximal
```{r plt_pair_ipa_pr, fig.width=6, fig.height=6}
g <- df_targets %>%
    ggplot(aes(x=pct_ipa, y=pct_proximal, color=gene_set, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x="Percent UMIs from IPA isoforms",
         y="Percent UMIs from proximal isoforms",
         color="Gene Set")
    
ggplotly(g, tooltip=c("text", "x", "y", "gene_set", "n_cells"))
```


```{r plt_pair_ipa_pr_static, fig.width=6, fig.height=6}
df_targets %>%
    ggplot(aes(x=pct_ipa, y=pct_proximal, color=gene_set, text=target_gene)) +
    geom_point(data=filter(df_targets, gene_set == 'other'), size=0.5) +
    geom_point(data=filter(df_targets, gene_set == 'non-targeting'), size=0.5) +
    geom_point(data=filter(df_targets, is_known), size=3) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    labs(x="Percent UMIs from IPA isoforms",
         y="Percent UMIs from proximal isoforms",
         color="Gene Set")
```

### IPA vs Distal
```{r plt_pair_ipa_ds, fig.width=6, fig.height=6}
g <- df_targets %>%
    ggplot(aes(x=pct_ipa, y=pct_distal, color=gene_set, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x="Percent UMIs from IPA isoforms",
         y="Percent UMIs from distal isoforms",
         color="Gene Set")
    
ggplotly(g, tooltip=c("text", "x", "y", "gene_set", "n_cells"))
```

```{r plt_pair_ipa_ds_static, fig.width=6, fig.height=6}
df_targets %>%
    ggplot(aes(x=pct_ipa, y=pct_distal, color=gene_set, text=target_gene)) +
    geom_point(data=filter(df_targets, gene_set == 'other'), size=0.5) +
    geom_point(data=filter(df_targets, gene_set == 'non-targeting'), size=0.5) +
    geom_point(data=filter(df_targets, is_known), size=3) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    labs(x="Percent UMIs from IPA isoforms",
         y="Percent UMIs from distal isoforms",
         color="Gene Set")
```

### IPA vs MT
```{r plt_pair_ipa_mt, fig.width=6, fig.height=6}
g <- df_targets %>%
    ggplot(aes(x=pct_ipa, y=pct_mt, color=gene_set, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x="Percent UMIs from IPA isoforms",
         y="Percent UMIs from chrM",
         color="Gene Set")
    
ggplotly(g, tooltip=c("text", "x", "y", "gene_set", "n_cells"))
```

```{r plt_pair_ipa_mt_static, fig.width=6, fig.height=6}
df_targets %>%
    ggplot(aes(x=pct_ipa, y=pct_mt, color=gene_set, text=target_gene)) +
    geom_point(data=filter(df_targets, gene_set == 'other'), size=0.5) +
    geom_point(data=filter(df_targets, gene_set == 'non-targeting'), size=0.5) +
    geom_point(data=filter(df_targets, is_known), size=3) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    labs(x="Percent UMIs from IPA isoforms",
         y="Percent UMIs from chrM",
         color="Gene Set")
```

### Proximal vs MT
```{r plt_pair_proximal_mt, fig.width=6, fig.height=6}
g <- df_targets %>%
    ggplot(aes(x=pct_proximal, y=pct_mt, color=gene_set, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x="Percent UMIs from proximal isoforms",
         y="Percent UMIs from chrM",
         color="Gene Set")
    
ggplotly(g, tooltip=c("text", "x", "y", "gene_set", "n_cells"))
```

```{r plt_pair_pr_mt_static, fig.width=6, fig.height=6}
df_targets %>%
    ggplot(aes(x=pct_proximal, y=pct_mt, color=gene_set, text=target_gene)) +
    geom_point(data=filter(df_targets, gene_set == 'other'), size=0.5) +
    geom_point(data=filter(df_targets, gene_set == 'non-targeting'), size=0.5) +
    geom_point(data=filter(df_targets, is_known), size=3) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    labs(x="Percent UMIs from proximal isoforms",
         y="Percent UMIs from chrM",
         color="Gene Set")
```


### Distal vs MT
```{r plt_pair_distal_mt, fig.width=6, fig.height=6}
g <- df_targets %>%
    ggplot(aes(x=pct_distal, y=pct_mt, color=gene_set, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x="Percent UMIs from distal isoforms",
         y="Percent UMIs from chrM",
         color="Gene Set")
    
ggplotly(g, tooltip=c("text", "x", "y", "gene_set", "n_cells"))
```

```{r plt_pair_ds_mt_static, fig.width=6, fig.height=6}
df_targets %>%
    ggplot(aes(x=pct_distal, y=pct_mt, color=gene_set, text=target_gene)) +
    geom_point(data=filter(df_targets, gene_set == 'other'), size=0.5) +
    geom_point(data=filter(df_targets, gene_set == 'non-targeting'), size=0.5) +
    geom_point(data=filter(df_targets, is_known), size=3) +
    geom_rug(size=0.1) +
    scale_color_manual(values=COLOR_MAP) +
    scale_size_continuous(range=c(0.05, 3)) +
    scale_x_continuous(labels=scales::percent_format(accuracy=0.1)) +
    scale_y_continuous(labels=scales::percent_format(accuracy=0.1)) +
    theme_bw() +
    labs(x="Percent UMIs from distal isoforms",
         y="Percent UMIs from chrM",
         color="Gene Set")
```

# Conclusion

The data set appears consistent with being well-processed and coherent. The most 
interesting observation is the significant perturbation of distal -> proximal isoforms 
in the knockdowns of the NUDT21, CPSF6, and OGDOF1. It should be noted that OGDOF1 and
NUDT21 form a divergent gene pair, raising the possibility that CRISPRi targeting OGDOF1 
also impacts NUDT21 expression levels, and thus does not have any significant effect
on isoform usage itself. Examining heatmaps from both the KD6 and KD8 experiments
shows that in the OGDOF1 perturbations, NUDT21 expression was also reduced to a 
similar z-score (though slightly higher than a direct NUDT21 perturbation).

Additionally, both the NUDT21 and OGDOF1 perturbations resulted in similar upregulation
of CPSF6 levels, however, CPSF6 perturbation showed only upregulation of NUDT21,
with no change in OGDOF1 expression.

---

# Runtime Details
## Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```

## Conda Environment
```{bash comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
