---
title: "ELAVL1 SU-LU Plots - GWPS"
author: "Mervin M Fansler"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
---

# Purpose

In this exploratory analysis, we use abundance of ELAVL1 isoforms as a readout for 
shifts in isoform usage.

# Initialization

## Libraries
```{r libs, message=FALSE, warning=FALSE}
library(magrittr)
library(tidyverse)
library(Matrix)
library(plotly)
```

## Parameters
```{r set_params}
set.seed(20210818)
GENE="ELAVL1"
FILE_ANNOTS="metadata/kd8-gwps-cell-annots.csv.gz"
FILE_CTS_TOTAL="data/counts/kd8_gwps.cts_total_cell.Rds"
RDS_CTS_GENE=sprintf("data/counts/raw/kd8_gwps.cts_%s_cell.Rds", tolower(GENE))

LBL_TXS=c("ENST00000593807.1"="cts_su", 
          "ENST00000596459.5"="cts_mu1",
          "ENST00000407627.7-UTR-3701"="cts_mu2",
          "ENST00000596154.5"="cts_mu3",
          "ENST00000407627.7-UTR-617"="cts_mu4",
          "ENST00000407627.7"="cts_lu")
FILE_OUT=sprintf("tbl/%s_gwps_targets.xlsx", tolower(GENE))
```

# Data
## Loading
```{r load_data, message=FALSE}
cts_total_cell <- readRDS(FILE_CTS_TOTAL)
cts_gene_cell <- readRDS(RDS_CTS_GENE)
df_cells <- read_csv(FILE_ANNOTS, col_types='cfiiicccccdddd')
```

## Preprocessing

```{r prepare_data}
## ensure cell order assumption
stopifnot(all(colnames(cts_gene_cell) == names(cts_total_cell)))

## sparse model matrix mapping cells to targets
M_cell_target <- df_cells %>% 
    select(cell_id, sgID_AB) %>% deframe() %>%
    fac2sparse() %>% t() %>%
    `[`(names(cts_total_cell),)

df_total <- drop(cts_total_cell %*% M_cell_target) %>%
    enframe(name="sgID_AB", value="cts_total")

df_gene <- t(cts_gene_cell %*% M_cell_target) %>%
    `colnames<-`(LBL_TXS[colnames(.)]) %>%
    as.matrix %>% as_tibble(rownames="sgID_AB")

df_targets <- df_cells %>% 
    group_by(target_gene, target_gene_id, sgID_AB) %>%
    summarize(n_cells=dplyr::n(), .groups="drop") %>%
    left_join(df_gene, by="sgID_AB") %>% 
    left_join(df_total, by="sgID_AB") %>%
    mutate(across(starts_with("cts"), ~ 1e6*.x/cts_total, .names="tpm_{.col}")) %>%
    rename_with(~ str_remove(.x, "cts_"), starts_with("tpm"))
```

## Export Table
```{r export}
writexl::write_xlsx(df_targets, FILE_OUT)
```

# Target Analysis
We will explore whether any perturbations have effects on `r GENE` isoform expression.

### `r GENE` SU vs MU1
```{r plt_pair_su_mu1, fig.width=8, fig.height=8}
g <- df_targets %>%
    mutate(is_nt=target_gene == "non-targeting") %>%
    ggplot(aes(x=tpm_su, y=tpm_mu1, color=is_nt, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=c("black", "red")) +
    scale_size_continuous(range=c(0.05, 3)) +
    #scale_size_manual(values=c(0.2, 0.4)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x=sprintf("TPM - %s SU", GENE),
         y=sprintf("TPM - %s MU1", GENE))

ggplotly(g, tooltip=c("text", "x", "y", "n_cells"))
```

Note that this has NUDT21 et al. in a swapped position. This first annotated isoform
comes from GENCODE and was not supported by any cell types in the HCL data; it lies
upstream of the stop codon and is rather lowly expressed. Suspicious.

### `r GENE` MU1 vs MU2
```{r plt_pair_mu1_mu2, fig.width=8, fig.height=8}
g <- df_targets %>%
    mutate(is_nt=target_gene == "non-targeting") %>%
    ggplot(aes(x=tpm_mu1, y=tpm_mu2, color=is_nt, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=c("black", "red")) +
    scale_size_continuous(range=c(0.05, 3)) +
    #scale_size_manual(values=c(0.2, 0.4)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x=sprintf("TPM - %s MU1", GENE),
         y=sprintf("TPM - %s MU2", GENE))

ggplotly(g, tooltip=c("text", "x", "y", "n_cells"))
```


### `r GENE` MU2 vs MU3
```{r plt_pair_mu2_mu3, fig.width=8, fig.height=8}
g <- df_targets %>%
    mutate(is_nt=target_gene == "non-targeting") %>%
    ggplot(aes(x=log10(1+tpm_mu2), y=log10(1+tpm_mu3), color=is_nt, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=c("black", "red")) +
    scale_size_continuous(range=c(0.05, 3)) +
    #scale_size_manual(values=c(0.2, 0.4)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x=sprintf("log10(TPM+1)- %s MU2", GENE),
         y=sprintf("log10(TPM+1) - %s MU3", GENE))

ggplotly(g, tooltip=c("text", "x", "y", "n_cells"))
```

Note that MU3 is from GENCODE and had no support in the HCL.

### `r GENE` MU2 vs MU4
```{r plt_pair_mu2_mu4, fig.width=8, fig.height=8}
g <- df_targets %>%
    mutate(is_nt=target_gene == "non-targeting") %>%
    ggplot(aes(x=log10(1+tpm_mu2), y=log10(1+tpm_mu4), color=is_nt, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=c("black", "red")) +
    scale_size_continuous(range=c(0.05, 3)) +
    #scale_size_manual(values=c(0.2, 0.4)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x=sprintf("log10(TPM+1) - %s MU2", GENE),
         y=sprintf("log10(TPM+1) - %s MU4", GENE))

ggplotly(g, tooltip=c("text", "x", "y", "n_cells"))
```

### `r GENE` MU4 vs LU
```{r plt_pair_mu4_lu, fig.width=8, fig.height=8}
g <- df_targets %>%
    mutate(is_nt=target_gene == "non-targeting") %>%
    ggplot(aes(x=log10(1+tpm_mu4), y=log10(1+tpm_lu), color=is_nt, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=c("black", "red")) +
    scale_size_continuous(range=c(0.05, 3)) +
    #scale_size_manual(values=c(0.2, 0.4)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x=sprintf("log10(TPM+1) - %s MU4", GENE),
         y=sprintf("log10(TPM+1) - %s LU", GENE))

ggplotly(g, tooltip=c("text", "x", "y", "n_cells"))
```


### `r GENE` MU2 vs LU
```{r plt_pair_mu2_lu, fig.width=8, fig.height=8}
g <- df_targets %>%
    mutate(is_nt=target_gene == "non-targeting") %>%
    ggplot(aes(x=log10(1+tpm_mu2), y=log10(1+tpm_lu), color=is_nt, text=target_gene)) +
    geom_point(aes(size=n_cells), pch=16) +
    geom_rug(size=0.1) +
    scale_color_manual(values=c("black", "red")) +
    scale_size_continuous(range=c(0.05, 3)) +
    #scale_size_manual(values=c(0.2, 0.4)) +
    theme_bw() +
    theme(legend.position="none") +
    labs(x=sprintf("log10(TPM+1) - %s MU2", GENE),
         y=sprintf("log10(TPM+1) - %s LU", GENE))

ggplotly(g, tooltip=c("text", "x", "y", "n_cells"))
```

Here we see **FIP1L1** stand out as factor whose knockdown increase long abundance.

---

# Runtime Details
## Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```

## Conda Environment
```{bash comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
